File: aoc-py/solutions/day20.py
from __future__ import annotations
"""
Day 20: Pulse Propagation

https://adventofcode.com/2023/day/20
"""
__all__ = ('Day20',)

from typing import ClassVar
from itertools import count
from math import lcm

from ..solution import Solution

class Destination:
    """Represents a destination when processing the button presses and signals"""
    __slots__ = ('target', 'pulse', 'name')

    def __init__(self, target: str, pulse: bool, *, name: str = 'broadcaster') -> None:
        self.name = name
        self.target = target
        self.pulse = pulse

    def __repr__(self) -> str:
        return f'<{self.__class__.__name__} name={self.name!r} target={self.target!r} pulse={self.pulse}>'

class Module:
    """Base Module Class"""
    __slots__ = ('name', 'outputs')

    def __init__(self, name: str, outputs: list[str]) -> None:
        self.name = name
        self.outputs = outputs

    def __repr__(self) -> str:
        return f'<{self.__class__.__name__} name={self.name!r}>'

class Flipper(Module):
    """Flipper type module, a module with a prefix of '%'

    Stores it 'on' or 'off' status in `self.status`
    """
    __slots__ = ('name', 'outputs', 'status')

    def __init__(self, name: str, outputs: list[str]) -> None:
        super().__init__(name, outputs)

        # on or off
        self.status = False

    def __repr__(self) -> str:
        return f'<{self.__class__.__name__} name={self.name!r}> status={self.status}>'

class Conjunction(Module):
    """Conjunction type module, a module with a prefix of '&'

    Stores the most recent pulses received from each of their connected input modules
    inside `self.memory` as a mapping of `module_name -> pulse`
    """
    __slots__ = ('name', 'outputs', 'memory')

    def __init__(self, name: str, outputs: list[str]) -> None:
        super().__init__(name, outputs)

        self.memory: dict[str, bool] = {}

    def __repr__(self) -> str:
        return f'<{self.__class__.__name__} name={self.name!r}> memory={self.memory}>'

class Day20(Solution):
    NAME: ClassVar[str] = 'Pulse Propagation'

    def _parse_input(self, inp: str) -> tuple[dict[str, Module], list[str]]:
        """Parses the input

        returns a mapping of modules and a list of our starting targets
        """
        modules: dict[str, Module] = {}
        broadcast_targets: list[str] = []

        for line in inp.splitlines():
            name, targets = line.split('->')
            name = name.strip()
            targets = [target.strip() for target in targets.split(',')]

            if name == 'broadcaster':
                # unique broadcaster module, our starting point
                broadcast_targets = targets
            else:
                match name[0], name[1:]:
                    case '%', name:
                        modules[name] = Flipper(name, targets)
                    case '&', name:
                        modules[name] = Conjunction(name, targets)

        # grab all modules: treat as input modules
        for name, module in modules.items():
            # grab each input module's outputs
            # check if it's a conjunction module (remember's its input modules)
            for output in module.outputs:
                if (
                    (module := modules.get(output))
                    and isinstance(module, Conjunction)
                ):
                    # fills all conjunction modules with their connected input modules
                    # defaulting to false to begin
                    module.memory[name] = False
        return modules, broadcast_targets

    def _run_modules(self, destinations: list[Destination], module: Module, source: str, pulse: bool) -> None:
        """Runs a single iteration of all the modules' pulse sending process"""

        if isinstance(module, Flipper) and not pulse:
            module.status = not module.status
            pulse_to_send = module.status
        elif isinstance(module, Conjunction):
            module.memory[source] = pulse
            # Sends low pulse if all remembered pulses are high otherwise sends a high pulse
            pulse_to_send = not all(module.memory.values())
        else:
            return

        for output in module.outputs:
            destinations.append(
                Destination(output, pulse_to_send, name=module.name)
            )

    def part_one(self, inp: str) -> int:
        modules, broadcast_targets = self._parse_input(inp)

        n_low = 0
        n_high = 0
        for _ in range(1000):
            n_low += 1
            destinations = [
                Destination(target, pulse=False)
                for target in broadcast_targets
            ]
            while destinations:
                source = destinations.pop(0)

                if source.pulse:
                    n_high += 1
                else:
                    n_low += 1

                if module := modules.get(source.target):
                    self._run_modules(destinations, module, source.name, source.pulse)
        return n_low * n_high

    def part_two(self, inp: str) -> int:
        modules, broadcast_targets = self._parse_input(inp)

        # finds the module that has an output into 'rx'
        rx_feeder = next(filter(
            lambda name: 'rx' in modules[name].outputs,
            modules,
        ))
        # backtrack and find all modules that have an output that is the feeder into 'rx'
        seen = {
            name: False
            for name, module in modules.items()
            if rx_feeder in module.outputs
        }
        # contains all modules in which the module outputs to the module that feeds into 'rx'
        # maps to the amount of button presses each one takes to complete
        press_amounts = {}

        for n_presses in count(start=1):
            destinations = [
                Destination(target, pulse=False)
                for target in broadcast_targets
            ]
            while destinations:
                source = destinations.pop(0)

                if module := modules.get(source.target):
                    if module.name == rx_feeder and source.pulse:
                        seen[name := source.name] = True

                        if name not in press_amounts:
                            press_amounts[name] = n_presses

                        if all(seen.values()):
                            # we have checked all the modules that feed into the module
                            # we can then return the result
                            # lowest common multiple of all buttonn press amounts of the modules that feed into the feeder of 'rx'
                            return lcm(*press_amounts.values())
                    self._run_modules(destinations, module, source.name, source.pulse)
        raise ValueError('No solution found')

    def run(self, inp: str) -> None:
        print('Part 1:', p1 := self.part_one(inp))
        print('Part 2:', p2 := self.part_two(inp))

        assert p1 == 680278040
        assert p2 == 243548140870057

File: src/bin/day20.rs
#![feature(map_try_insert)]

//! Day 20: Pulse Propagation
//!
//! <https://adventofcode.com/2023/day/20>
use std::{
    collections::{VecDeque, HashMap},
    fmt::Display,
};
use aoc_2023::{Solution, lcm};

#[derive(Debug, Clone)]
struct Destination<'a> {
    name: &'a str,
    target: String,
    pulse: bool,
}

#[derive(Debug, Clone)]
enum Module<'a> {
    Flipper {
        name: &'a str,
        outputs: Vec<String>,
        status: bool,
    },
    Conjunction {
        name: &'a str,
        outputs: Vec<String>,
        memory: HashMap<String, bool>,
    }
}

impl<'a> Module<'a> {
    #[must_use]
    fn new(raw: &'a str, outputs: Vec<String>) -> Self {
        match raw
            .split_at(1)
        {
            ("%", name) => Self::Flipper {
                name, outputs,
                status: false,
            },
            ("&", name) => Self::Conjunction {
                name, outputs,
                memory: HashMap::new(),
            },
            _ => panic!("Invalid module prefix"),
        }
    }

    #[inline]
    #[must_use]
    const fn name(&self) -> &'a str {
        match self {
            Self::Flipper { name, .. }
            | Self::Conjunction { name, .. } => name,
        }
    }

    #[inline]
    #[must_use]
    const fn outputs(&self) -> &Vec<String> {
        match self {
            Self::Flipper { outputs, .. }
            | Self::Conjunction { outputs, .. } => outputs,
        }
    }
}

pub struct Day20;

impl Day20 {
    #[must_use]
    fn parse_input<T>(inp: &T) -> (HashMap<String, Module<'_>>, Vec<String>)
    where
        T: AsRef<str>,
    {
        let mut modules = HashMap::new();
        let mut broadcast_targets = Vec::new();

        for line in inp
            .as_ref()
            .lines()
        {
            let (name, targets) = line
                .split_once("->")
                .unwrap();
            let targets = targets
                .split(',')
                .map(|s| s.trim().to_string())
                .collect::<Vec<String>>();

            match name.trim() {
                "broadcaster" => broadcast_targets = targets,
                name => {
                    let module = Module::new(name, targets);
                    modules.insert(
                        module.name().to_string(),
                        module,
                    );
                }
            }
        }

        for (name, module) in modules
            .clone()
        {
            for output in module
                .outputs()
            {
                if let Some(Module::Conjunction {
                    memory, ..
                }) = modules.get_mut(output) {
                    memory.insert(
                        name.to_string(),
                        false
                    );
                }
            }
        }
        (modules, broadcast_targets)
    }

    fn run_modules<'a>(
        destinations: &mut VecDeque<Destination<'a>>,
        mut module: &mut Module<'a>,
        source: &'a str,
        pulse: bool,
    ) {
        let pulse_to_send = match (&mut module, pulse) {
            (Module::Flipper { status, .. }, false) => {
                *status = !*status;
                *status
            },
            (Module::Conjunction { memory, .. }, pulse) => {
                memory.insert(source.to_string(), pulse);
                !memory
                    .values()
                    .all(|x| *x)
            },
            _ => return,
        };

        for output in module.outputs() {
            destinations.push_back(
                Destination {
                    target: output.to_string(),
                    pulse: pulse_to_send,
                    name: module.name(),
                }
            );
        }
    }

    #[inline]
    #[must_use]
    fn get_destinations<'a, T>(broadcast_targets: T) -> VecDeque<Destination<'a>>
    where
        T: AsRef<[String]>,
    {
        broadcast_targets
            .as_ref()
            .iter()
            .map(|target| Destination {
                target: target.to_string(),
                pulse: false,
                name: "broadcaster",
            })
            .collect::<VecDeque<Destination>>()
    }
}

impl Solution for Day20 {
    const NAME: &'static str = "Pulse Propagation";

    fn part_one<T: Display>(&self, inp: T) -> usize {
        let inp = inp.to_string();
        let (mut modules, broadcast_targets) =
            Self::parse_input(&inp);

        let mut n_low = 0;
        let mut n_high = 0;

        for _ in 0..1000 {
            n_low += 1;
            let mut destinations =
                Self::get_destinations(&broadcast_targets);

            while let Some(source) = destinations.pop_front() {
                if source.pulse {
                    n_high += 1;
                } else {
                    n_low += 1;
                }

                if let Some(module) = modules.get_mut(&source.target) {
                    Self::run_modules(
                        &mut destinations,
                        module,
                        source.name,
                        source.pulse
                    );
                }
            }
        }
        n_low * n_high
    }

    /// # Panics
    ///
    /// If the module that feeds into 'rx' is not found
    fn part_two<T: Display>(&self, inp: T) -> usize {
        let inp = inp.to_string();
        let (mut modules, broadcast_targets) =
            Self::parse_input(&inp);

        let rx_feeder = modules
            .iter()
            .find_map(|(name, module)| module
                .outputs()
                .contains(&"rx".to_string())
                .then_some(name)
            )
            .unwrap()
            .to_string();

        let mut seen = modules
            .iter()
            .filter_map(|(name, module)|
                module
                    .outputs()
                    .contains(&rx_feeder)
                    .then_some((
                        name.to_string(),
                        false,
                    ))
            )
            .collect::<HashMap<String, bool>>();

        let mut press_amounts = HashMap::new();

        for n_presses in 1.. {
            let mut destinations =
                Self::get_destinations(&broadcast_targets);

            while let Some(source) = destinations.pop_front() {
                if let Some(module) = modules.get_mut(&source.target) {
                    if module.name() == rx_feeder && source.pulse {
                        seen.insert(
                            source.name.to_string(),
                            true,
                        );
                        press_amounts
                            .try_insert(source.name, n_presses)
                            .ok();

                        if seen
                            .values()
                            .all(|b| *b)
                        {
                            return lcm(press_amounts
                                .values()
                                .copied()
                            );
                        }
                    }
                    Self::run_modules(
                        &mut destinations,
                        module,
                        source.name,
                        source.pulse,
                    );
                }
            }
        }
        unreachable!()
    }

    fn run(&self, inp: String) {
        let p1 = self.part_one(&inp);
        let p2 = self.part_two(&inp);

        println!("Part 1: {p1}");
        println!("Part 2: {p2}");

        assert_eq!(p1, 680_278_040);
        assert_eq!(p2, 243_548_140_870_057);
    }
}

fn main() {
    aoc_2023::run_day(20, &Day20);
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test() { main(); }
}